### PR TITLE
Add support for BAP reporting endpoint

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/14315

## Submitter Checklist:

- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)
- [ ] Requested a security/privacy review as needed

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by kylehickinson — APPROVED** [review]
iOS++

File: vendor/bat-native-ledger/src/bat/ledger/internal/promotion/bap_reporter.cc [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  // First, get the user's unspent BAP tokens
+  ledger_->database()->GetSpendableUnblindedTokensByBatchTypes(
+      {type::CredsBatchType::PROMOTION},
+      std::bind(&BAPReporter::OnUnblindedTokens, this, _1));
+}
+
+void BAPReporter::OnUnblindedTokens(
+    std::vector<type::UnblindedTokenPtr> tokens) {
+  double amount = 0;
+  for (auto& token : tokens)
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/promotion/bap_reporter.cc:60]: Nit: Can this be `const`?

File: vendor/bat-native-ledger/src/bat/ledger/internal/promotion/bap_reporter_endpoint.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    case net::HTTP_BAD_REQUEST: {
+      BLOG(0, "Invalid request");
+      break;
+    }
+    case net::HTTP_INTERNAL_SERVER_ERROR: {
+      BLOG(0, "Internal server error");
+      break;
+    }
+    case net::HTTP_OK:
+    case net::HTTP_CONFLICT: {
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/promotion/bap_reporter_endpoint.cc:78]: What would `HTTP_CONFLICT` signal in this scenario?

Comment (emerick) [reply vendor/bat-native-ledger/src/bat/ledger/internal/promotion/bap_reporter_endpoint.cc:78]: Nevermind, it looks like it just means that it's already been called.

**Review by emerick — APPROVED** [review]
LGTM, just had a minor nit.

**Comment by zenparsing** [comment]
CI post-init failed on unrelated `audit-deps` error.

**Comment by LaurenWags** [comment]
Verification in progress with

```
Brave	1.23.20 Chromium: 89.0.4389.72 (Official Build) nightly (x86_64)
Revision	3f345f156bfd157bd1bea06310e55f3fb2490359-refs/branch-heads/4389@{#1393}
OS	macOS Version 10.15.7 (Build 19H512)
```




<details>
<summary>Test Case 4 - Server returns 400, reporting goes into retries</summary>

1. Clean profile (`--args --enable-logging=stderr --vmodule="*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*/bat-native-confirmations/*"=6,"*/brave_ads/*"=6,"*/brave_user_model/*"=6 --brave-ads-staging --rewards=staging=true,persist-logs=true`)
2. Enable Rewards
3. Go to brave://rewards-internals, confirm BAP retry is scheduled (can also check in terminal)
4. Claim UGP grant
5. Close browser
6. Set up Charles Proxy rule for the server to return 400 for 200 or 409
7. Restart with same flags
8. Watch logs for scheduled retries

```
[ REQUEST ]
> URL: https://grant.rewards.bravesoftware.com/v1/promotions/report-bap
> Method: UrlMethod::POST
> Content: {"amount":30.0}
> Content Type: application/json; charset=utf-8
> Header digest: SHA-256=Z/frhwyMcI3GeHys2lrUHmtSKj/pkTWttptg5TAw+6o=
> Header signature: keyId="f0d10740-cfe1-4f6c-b6dc-282eb3be5ca2",algorithm="ed25519",headers="digest (request-target)",signature="kqak8ZQofnjW5pi61Q/l4ZK4cx9OqlPJeAbv2qBWQwbBEI2UAf+PnRuD6hSykqM2RcX4iCRYMR3kkxygF9fpCg=="
> Header accept: application/json

[ RESPONSE - OnFetchCompleted ]
> Url: https://grant.rewards.bravesoftware.com/v1/promotions/report-bap
> Result: Failure
> HTTP Code: 400
> Body: {"reportBapId":"360f5c86-8228-4e5e-875c-f72cc86f4020"}

[158:775:0305/100257.222750:INFO:bap_reporter_endpoint.cc(70)] Invalid request
[158:775:0305/100257.222796:VERBOSE1:bap_reporter.cc(108)] BAP reporting failed - rescheduling in 717 s
```

```
[ REQUEST ]
> URL: https://grant.rewards.bravesoftware.com/v1/promotions/report-bap
> Method: UrlMethod::POST
> Content: {"amount":30.0}
> Content Type: application/json; charset=utf-8
> Header digest: SHA-256=Z/frhwyMcI3GeHys2lrUHmtSKj/pkTWttptg5TAw+6o=
> Header signature: keyId="f0d10740-cfe1-4f6c-b6dc-282eb3be5ca2",algorithm="ed25519",headers="digest (request-target)",signature="kqak8ZQofnjW5pi61Q/l4ZK4cx9OqlPJeAbv2qBWQwbBEI2UAf+PnRuD6hSykqM2RcX4iCRYMR3kkxygF9fpCg=="
> Header accept: application/json
[158:775:0305/101454.679877:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnFetchCompleted ]
> Url: https://grant.rewards.bravesoftware.com/v1/promotions/report-bap
> Result: Failure
> HTTP Code: 400
> Body: {"message":"Error inserting bap report, paymentId already reported: unable to record BAP report","code":409}

[158:775:0305/101454.679972:INFO:bap_reporter_endpoint.cc(70)] Invalid request
[158:775:0305/101454.680028:VERBOSE1:bap_reporter.cc(108)] BAP reporting failed - rescheduling in 683 s
```

```
[ REQUEST ]
> URL: https://grant.rewards.bravesoftware.com/v1/promotions/report-bap
> Method: UrlMethod::POST
> Content: {"amount":30.0}
> Content Type: application/json; charset=utf-8
> Header digest: SHA-256=Z/frhwyMcI3GeHys2lrUHmtSKj/pkTWttptg5TAw+6o=
> Header signature: keyId="f0d10740-cfe1-4f6c-b6dc-282eb3be5ca2",algorithm="ed25519",headers="digest (request-target)",signature="kqak8ZQofnjW5pi61Q/l4ZK4cx9OqlPJeAbv2qBWQwbBEI2UAf+PnRuD6hSykqM2RcX4iCRYMR3kkxygF9fpCg=="
> Header accept: application/json
[158:775:0305/102618.140851:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnFetchCompleted ]
> Url: https://grant.rewards.bravesoftware.com/v1/promotions/report-bap
> Result: Failure
> HTTP Code: 400
> Body: {"message":"Error inserting bap report, paymentId already reported: unable to record BAP report","code":409}

[158:775:0305/102618.140934:INFO:bap_reporter_endpoint.cc(70)] Invalid request
[158:775:0305/102618.140971:VERBOSE1:bap_reporter.cc(108)] BAP reporting failed - rescheduling in 4832 s
```

Note - geometric timer w/ randomness is used so I am unable to confirm that it backs off exactly as listed in https://bravesoftware.slack.com/archives/CB0KCRCQN/p1614191899060000?thread_ts=1614111570.049100&cid=CB0KCRCQN, but I did confirm with @zenparsing that it looked correct

</details>


<details>
<summary>Test Case 8 - Server returns 409, reporting does not retry</summary>

1. Clean profile (`--args --enable-logging=stderr --vmodule="*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*/bat-native-confirmations/*"=6,"*/brave_ads/*"=6,"*/brave_user_model/*"=6 --brave-ads-staging --rewards=staging=true,persist-logs=true`)
2. Enable Rewards
3. Go to brave://rewards-internals, confirm BAP retry is scheduled (can also check in terminal)
4. Claim UGP grant
5. Close the browser
6. Restart the browser w/ same flags
7. Look at logs in terminal or on brave://rewards-internals, confirm BAP is successfully sent to the server
8. Close browser
9. Update `brave.rewards.bap_reported` in Preferences file to be "false" to indicate that BAP was not sent
10. Restart the browser w/ same flags
11. Watch logs for 409 response

```
[ REQUEST ]
> URL: https://grant.rewards.bravesoftware.com/v1/promotions/report-bap
> Method: UrlMethod::POST
> Content: {"amount":30.0}
> Content Type: application/json; charset=utf-8
> Header digest: SHA-256=Z/frhwyMcI3GeHys2lrUHmtSKj/pkTWttptg5TAw+6o=
> Header signature: keyId="85e65a56-80f0-4b05-ba02-34eccf13b934",algorithm="ed25519",headers="digest (request-target)",signature="xdCCaoytAb0czdhVKGSQSbuTN+JvQn1ayLo9WLvmEehBy4W60QkScXCv4fL0Jxvk/jrH2k988h41by9gYj0WDQ=="
> Header accept: application/json
[ RESPONSE - OnFetchCompleted ]
> Url: https://grant.rewards.bravesoftware.com/v1/promotions/report-bap
> Result: Failure
> HTTP Code: 409
> Body: {"message":"Error inserting bap report, paymentId already reported: unable to record BAP report","code":409}
[99655:775:0305/094031.423218:VERBOSE1:bap_reporter.cc(82)] BAP successsfully reported
```

12. Close and relaunch w/ same flags
13. Confirm BAP is not retried

</details>

**Pull request merged**